modules/javafx.base/src/main/java/com/sun/javafx/binding/ListenerList.java

```diff
@@ -0,0 +1,195 @@
+     * all intended listeners (up to the maximum set by the higher level loop)
+     * were called and so there may be confusion as to the current value is now
+     * for some listeners.
+     *
+     * An abort may indicate a problem with multiple listeners changing values
+     * that are not converging to a mutually agreed value.
+     */
+    private static final int NESTED_NOTIFICATION_ABORTED = -2;
+
+    /**
+     * Indicates a nested notification completed normally. After a nested
+     * notification (aborted or not), the current value is re-read so listeners
+     * not notified yet at the current level receive the newly updated current
+     * value.
+     */
+    private static final int NESTED_NOTIFICATION_COMPLETED = -1;
+
+    /**
+     * This field is only used during notifications, and only relevant
+     * when nested notifications occur. It is used for communicating
+     * information between the different nesting levels. To deeper
+     * nesting levels it contains the number of listeners that have
+     * been notified in higher level loops, while deeper nesting levels
+     * communicate to higher level loops whether a nested notification
+     * actually occurred if it completed normally or was aborted early.<p>
+     *
+     * When its value is zero or positive, it indicates the number of
+     * listeners notified in a higher level loop (minus one), while the constants
+     * {@link #NESTED_NOTIFICATION_ABORTED} and {@link #NESTED_NOTIFICATION_COMPLETED}
+     * indicate to a higher level loop that a nested notification occurred,
+     * requiring, for example, a refresh of the current value and a new equals
+     * check.
+     */
+    private int progress;
+
+    /**
+     * Creates a new instance with two listeners.
+     *
+     * @param listener1 a listener, cannot be {@code null}
+     * @param listener2 a listener, cannot be {@code null}
+     * @throws NullPointerException when any parameter is {@code null}
+     */
+    public ListenerList(Object listener1, Object listener2) {
+        super(listener1, listener2);
+    }
+
+    /**
+     * Notifies all listeners using the given observable value as source.
+     *
+     * @param observableValue an {@link ObservableValue}, cannot be {@code null}
+     * @param oldValue the value held by the observable before it was changed, can be {@code null}
+     * @return {@code true} if the listener list is not locked, and it was modified during
+     *     notification otherwise {@code false}
+     */
+    public boolean notifyListeners(ObservableValue<? extends T> observableValue, T oldValue) {
+        boolean wasLocked = isLocked();
+
+        if (!wasLocked) {
+            lock();
+        }
+
+        int initialProgress = progress;  // save as it will be modified soon
+        int invalidationListenersSize = invalidationListenersSize();
+        int maxInvalidations = wasLocked ? Math.min(initialProgress + 1, invalidationListenersSize) : invalidationListenersSize;
```

Comment:
You could break up the very long lines:

```suggestion
        int maxInvalidations = wasLocked
            ? Math.min(initialProgress + 1, invalidationListenersSize)
            : invalidationListenersSize;
```